lib/sensors/accelerometer.cpp

```diff
@@ -0,0 +1,140 @@
+#include "accelerometer.hpp"
+
+namespace hyped::sensors {
+
+Accelerometer::Accelerometer(core::ILogger &logger,
+                             io::HardwareI2c &i2c,
+                             const std::uint8_t channel)
+    : logger_(logger),
+      i2c_(i2c),
+      channel_(channel)
+{
+}
+
+Accelerometer::~Accelerometer()
+{
+}
+
+std::uint8_t Accelerometer::getChannel()
+{
+  return channel_;
+}
+
+std::optional<std::int16_t> Accelerometer::getRawAcceleration(const Axis axis)
+{
+  std::uint8_t low_byte_address, high_byte_address;
+  setRegisterAddressFromAxis(axis, &low_byte_address, &high_byte_address);
+  const auto low_byte = i2c_.readByte(kDeviceAddress, low_byte_address);
+  if (!low_byte) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to read the low byte for acceleration along the %s",
+                AxisStrings[axis]);
+    return std::nullopt;
+  }
+  const auto high_byte = i2c_.readByte(kDeviceAddress, high_byte_address);
+  if (!high_byte) {
+    logger_.log(core::LogLevel::kFatal,
+                "Failed to read the low byte for acceleration along the %s",
+                AxisStrings[axis]);
+    return std::nullopt;
+  }
+  const auto raw_acceleration
+    = static_cast<std::int16_t>((high_byte.value() << 8) | low_byte.value());
+  return raw_acceleration;
+}
+
+void Accelerometer::setRegisterAddressFromAxis(const Axis axis,
+                                               std::uint8_t *low_byte_address,
+                                               std::uint8_t *high_byte_address)
+{
+  switch (axis) {
+    case Axis::x:
+      *low_byte_address  = kXOutLow;
+      *high_byte_address = kXOutHigh;
+      break;
+    case Axis::y:
+      *low_byte_address  = kYOutLow;
+      *high_byte_address = kYOutHigh;
+      break;
+    case Axis::z:
+      *low_byte_address  = kZOutLow;
+      *high_byte_address = kZOutHigh;
+      break;
+  }
+}
+
+std::int16_t Accelerometer::getAccelerationFromRaw(const std::int16_t rawAcceleration)
+{
+  // these values come from the data sheet. Don't change them.
+  std::int16_t acceleration
+    = static_cast<std::int16_t>((static_cast<int32_t>(rawAcceleration) * 488) / 1000);
+  return acceleration;
+}
+
+// TODOlater: current settings of the accelerometer make it read in +-16g but with high noise. Check
+// whether other configuration could be better
+std::optional<core::RawAccelerationData> Accelerometer::read()
+{
+  // check to see if the values are ready to be read
+  const auto data_ready = i2c_.readByte(kDeviceAddress, kDataReady);
+  if (!data_ready) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read acceleration data");
+    return std::nullopt;
+  }
+  // TODOlater: here the error is not that bad, hence the return value should indicate that
+  if (data_ready.value() % 2 == 0) {
+    logger_.log(core::LogLevel::kFatal, "Failed to read acceleration data as it is not ready");
```

Comment:
Does this need to get logged as `kFatal`?

Reply:
No, this is an issue I've brought up and want to fix in another PR (this one has been around much too long). Ideally it would be `kDebug` and assuming this runs on a ROS node - the reaction to this "error" should be simply to try to read again till there is data (implementation mileage may vary*)

Reply:
Ok, sounds good. Might be a bit pedantic but is would `kDebug` really be the most suitable name for that behavior or would we be better off defining a new log level for this purpose i.e. `kRetry`/`kTempError` (basically something along the lines of "small oopsie but nothing too crashy-burny"? This is something that Navigation and the outlier detection system could potentially use too if we decide to add this? @SnickeyX ?

Reply:
I don't see this happening often enough (or ever) but if you can make use of it too then a `kWarn` makes sense since this WARN logs are usually for non-fatal conditions that should not majorly disrupt any processes. I could add this as part of the PR to fix this issue. What do you think @kshxtij?

Reply:
I agree, feel free to add it